hal/api/mbed_mem_trace.h

```diff
@@ -0,0 +1,138 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __MBED_MEM_TRACE_H__
+#define __MBED_MEM_TRACE_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <sys/types.h>
+
+/* Masks for tracer */
+#define MBED_MEM_TRACE_MALLOC           (1 << 0)
+#define MBED_MEM_TRACE_REALLOC          (1 << 1)
+#define MBED_MEM_TRACE_CALLOC           (1 << 2)
+#define MBED_MEM_TRACE_FREE             (1 << 3)
+#define MBED_MEM_TRACE_ALL              (MBED_MEM_TRACE_MALLOC | MBED_MEM_TRACE_REALLOC | MBED_MEM_TRACE_CALLOC | MBED_MEM_TRACE_FREE)
+#define MBED_MEM_TRACE_DISABLE          (0)
+
+/* Prefix for the output of the default tracer */
+#define MBED_MEM_DEFAULT_TRACER_PREFIX  "#"
+
+/**
+ * Type of the callback used by the memory tracer. This callback is called when a memory
+ * allocation operation (malloc, realloc, calloc, free) is called and tracing is enabled
+ * for that memory allocation function.
+ *
+ * @param op the ID of the operation (MBED_MEM_TRACE_MALLOC, MBED_MEM_TRACE_REALLOC,
+ *           MBED_MEM_TRACE_CALLOC or MBED_MEM_TRACE_FREE).
+ * @param res the result that the memory operation returned (NULL for 'free').
+ * @param caller the caller of the memory operation. Note that the value of 'caller' might be
+ *               unreliable.
+ *
+ * The rest of the parameters passed 'mbed_mem_trace_cb_t' are the same as the memory operations
+ * that triggered its call (see 'man malloc' for details):
+ *
+ * - for malloc: cb(MBED_MEM_TRACE_MALLOC, res, caller, size).
+ * - for realloc: cb(MBED_MEM_TRACE_REALLOC, res, caller, ptr, size).
+ * - for calloc: cb(MBED_MEM_TRACE_CALLOC, res, caller, nmemb, size).
+ * - for free: cb(MBED_MEM_TRACE_FREE, NULL, caller, ptr).
+ */
+typedef void (*mbed_mem_trace_cb_t)(uint8_t op, void *res, void* caller, ...);
+
+/**
+ * Set up for the memory tracer. This function can be called more than one time
+ * to set up different masks and/or callbacks.
+ *
+ * @param mask the tracer mask. It can be a bitwise combination of MBED_MEM_TRACE_MALLOC,
+ *             MBED_MEM_TRACE_REALLOC, MBED_MEM_TRACE_CALLOC, MBED_MEM_TRACE_FREE, or
+ *             MBED_MEM_TRACE_ALL to trace all memory operations, or MBED_MEM_TRACE_DISABLE
+ *             to disable tracing.
+ * @param cb the callback to call on each enabled memory operation.
+ */
+void mbed_mem_trace_setup(uint8_t mask, mbed_mem_trace_cb_t cb);
+
+/**
+ * Trace a call to 'malloc'.
+ * @param res the result of running 'malloc'.
+ * @param size the 'size' argument given to 'malloc'.
+ * @return 'res' (the first argument).
+ */
+void* mbed_mem_trace_malloc(void *res, size_t size);
```

Comment:
I admit that I didn't think much about optimizing these functions for speed, it seems like they won't be used when speed is an issue anyway.